### PR TITLE
Copy Elasticsearch configuration files before creating links

### PR DESCRIPTION
**Pull request by barkbay**
If Elasticsearch configuration files are copied before symbolic links are created, it is possible to run all the Elasticsearch containers with the following restricted `securityContext`

```yaml
    securityContext:
      allowPrivilegeEscalation: false
      capabilities:
        drop:
        - ALL
      privileged: false
      readOnlyRootFilesystem: true
```

Fixes https://github.com/elastic/cloud-on-k8s/issues/6126 

See my comment here if you want to review this PR: https://github.com/elastic/cloud-on-k8s/pull/6703#issuecomment-1519617415

### CODE REVIEW

**Comment by barkbay** [comment]
buildkite test this -f p=gke,s=8.7.0,s=7.17.8

**Comment by thbkrkr** [comment]
> buildkite test this -f p=gke,s=8.7.0,s=7.17.8

This will only run the e2e tests with `s=7.17.8`. When you want to test one variable with different values (here `s` the stack version), you need to declare this variable and its values in the `--mixed|-m` flag. 

- `buildkite test this --fixed p=gke --mixed s=8.7.0,s=7.17.8`
- `buildkite test this -f p=gke -m s=8.7.0,s=7.17.8` (with the short flags)

**Comment by barkbay** [comment]
* `TestVolumeEmptyDir` failed because this "feature" is itself broken: https://github.com/elastic/cloud-on-k8s/issues/6186. I added a function to detect, on a best effort basis, whether or not data are in a volume.
* `TestKibanaStandalone` also failed, but I'm not sure to understand why 🤷

**Comment by thbkrkr** [comment]
`TestKibanaStandalone` looks very similar to https://github.com/elastic/cloud-on-k8s/issues/6693 with all sub-tests failing as we don't have the end of the test execution in the log because the log stream was interrupted.

<details>

<summary>log</summary>

```json
{
    "Time": "2023-04-19T17:16:36.671150963Z",
    "Action": "output",
    "Package": "github.com/elastic/cloud-on-k8s/v2/test/e2e/kb",
    "Test": "TestKibanaStandalone/ES_HTTP_certificate_authority_should_be_set_and_deployed",
    "Output": "=== RUN   TestKibanaStandalone/ES_HTTP_certificate_authority_should_be_set_and_deployed\n"
}
{
    "log.level": "info",
    "@timestamp": "2023-04-19T17:30:05.540Z",
    "log.logger": "e2e-pdzcj",
    "message": "Log stream ended",
    "service.version": "0.0.0-SNAPSHOT+00000000",
    "service.type": "eck",
    "ecs.version": "1.4.0",
    "pod_name": "eck-e2e-pdzcj-cl86s"
}
{
    "log.level": "info",
    "@timestamp": "2023-04-19T17:30:05.540Z",
    "log.logger": "e2e-pdzcj",
    "message": "Streaming pod logs",
    "service.version": "0.0.0-SNAPSHOT+00000000",
    "service.type": "eck",
    "ecs.version": "1.4.0",
    "pod_name": "eck-e2e-pdzcj-cl86s"
}
```
</detail>

**Comment by barkbay** [comment]
buildkite test this --fixed p=gke --mixed s=8.7.0,s=7.17.8

**Comment by barkbay** [comment]
buildkite test this -f p=gke -m s=8.7.0,s=7.17.8

**Comment by barkbay** [comment]
Failures to investigate from the last e2e tests session:

image

I think this is because I added a check to validate that all the containers in the Elasticsearch `Pod` are running with the expected `securityContext`. That's probably not the case for Beat sidecars when stack monitoring is enabled.

**Comment by barkbay** [comment]
buildkite test this -f p=gke -m s=8.7.0,s=7.17.8

**Comment by barkbay** [comment]
buildkite test this -f p=gke -m s=8.7.0,s=7.17.8

**Comment by barkbay** [comment]
buildkite test this -f p=gke -m s=8.0.1,s=8.8.0-SNAPSHOT

**Comment by barkbay** [comment]
buildkite test this -f p=ocp,s=8.7.0

**Comment by barkbay** [comment]
Results from the last e2e test sessions:
* OCP4 🟢 
* `7.17.8`/`8.0.1`/`8.7.0` on GKE 🟢 
* `8.8.0-SNAPSHOT` 🟡 - 140/144 tests passed, `TestKibanaStandalone` failed. As for this [failure](https://github.com/elastic/cloud-on-k8s/pull/6703#issuecomment-1515892450), I don't think it is related to the changes in this PR.

**Review by thbkrkr — COMMENTED** [review]
LGTM, this is great!

Looks also like we're writing code specifically to handle #6186 but why not fix it first? We would no longer need to detect if the data directory is mounted in a volume as this is normally impossible, no? So we could call `WithContainersSecurityContext` after `stackmon.WithMonitoring`, to reuse the same code to set the SecurityContext for beats sidecars.

**Comment by barkbay** [comment]
> Looks also like we're writing code specifically to handle https://github.com/elastic/cloud-on-k8s/issues/6186 but why not fix it first?

At first I thought this code was specifically handling #6186. But I've been wondering if, in general, we should not set `readOnlyRootFilesystem` to `true` only if we are pretty confident in the fact that the data are written in a volume, not in the container. I'm thinking about users that will upgrade to `2.8.0`: if we set `readOnlyRootFilesystem` to true too "aggressively" we may end up with issues during upgrades if Elasticsearch is not configured properly. That's why I decided to set `readOnlyRootFilesystem` to `true` only if the default data volume is mounted as expected. If a user has a more involved storage configuration, with different volumes, it would have to test this flag manually first.

I have to admit this is still brittle, as there are other ways to mis-configure `Elasticsearch` with the default volume, ending up with data being written into the container. As you suggest, we could also always enable `readOnlyRootFilesystem` and let the user fix the configuration if the Pods crash, I'm trying to avoid that situation by keeping this code.

**Review by thbkrkr — APPROVED** [review]
> But I've been wondering if, in general, we should not set readOnlyRootFilesystem to true only if we are pretty confident in the fact that the data are written in a volume, not in the container. I'm thinking about users that will upgrade to 2.8.0: if we set readOnlyRootFilesystem to true too "aggressively" we may end up with issues during upgrades if Elasticsearch is not configured properly. 

👍 👍

**Review by pebrc — COMMENTED** [review]
Code LGTM I have not done many tests yet.

File: pkg/controller/elasticsearch/securitycontext/securitycontext.go [hunk 1]
```diff
@@ -0,0 +1,33 @@
...
+import (
+	corev1 "k8s.io/api/core/v1"
+	ptr "k8s.io/utils/pointer"
+
+	"github.com/elastic/cloud-on-k8s/v2/pkg/controller/common/version"
+)
+
+var (
+	// MinStackVersion is the minimum Stack version to use RunAsNonRoot with the Elasticsearch image.
+	MinStackVersion = version.MustParse("8.8.0-SNAPSHOT")
```
[the remark this answers is not in the document]
Comment (pebrc) [reply pkg/controller/elasticsearch/securitycontext/securitycontext.go:16]: We should maybe explain that this is due to the numeric user in the Docker image? and maybe even link to the ES issue

File: pkg/controller/elasticsearch/stackmon/sidecar.go [hunk 2]
```diff
@@ -49,11 +50,33 @@ func Metricbeat(ctx context.Context, client k8s.Client, es esv1.Elasticsearch) (
...
+		return stackmon.BeatSidecar{}, err
+	}
+	fileBeat.Container.SecurityContext = &corev1.SecurityContext{
+		Capabilities: &corev1.Capabilities{
+			Drop: []corev1.Capability{"ALL"},
+		},
+		Privileged:               pointer.Bool(false),
+		RunAsNonRoot:             pointer.Bool(true),
+		ReadOnlyRootFilesystem:   pointer.Bool(true),
+		AllowPrivilegeEscalation: pointer.Bool(false),
```
[the remark this answers is not in the document]
Comment (barkbay) [reply pkg/controller/elasticsearch/stackmon/sidecar.go]: 👍 done in https://github.com/elastic/cloud-on-k8s/pull/6703/commits/18818fc1fd95345e6acd9d8f1d1c3ecab99e104e

**Comment by barkbay** [comment]
I'll open an issue to update https://www.elastic.co/guide/en/cloud-on-k8s/current/k8s-security-context.html before the next release as it is definitely outdated.

**Pull request merged**